diff --git a/build/hemocell/CMakeLists.txt b/build/hemocell/CMakeLists.txt
index b238a35..b65c550 100644
--- a/build/hemocell/CMakeLists.txt
+++ b/build/hemocell/CMakeLists.txt
@@ -117,8 +117,6 @@ FILE(GLOB_RECURSE SRC_FILES_SRC "${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR}/IO/*.c
 SET(SRC_FILES ${SRC_FILES} ${SRC_FILES_SRC}) 
 FILE(GLOB_RECURSE SRC_FILES_SRC "${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR}/mechanics/*.cpp")
 SET(SRC_FILES ${SRC_FILES} ${SRC_FILES_SRC}) 
-FILE(GLOB_RECURSE SRC_FILES_SRC "${CMAKE_SOURCE_DIR}/${PALABOS_DIR}/src/*.cpp")
-SET(SRC_FILES ${SRC_FILES} ${SRC_FILES_SRC})
 
 INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR})
 INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR}/config)
@@ -128,13 +126,6 @@ INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR}/helper)
 INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR}/IO)
 INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR}/mechanics)
 INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR}/external/tinyxml2/)
-INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR}/palabos/src)
-INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/${HEMOCELL_BASE_DIR}/palabos/externalLibraries)
-
-#INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/${PALABOS_DIR}/externalLibraries)
-LIST(APPEND SRC_FILES ${CMAKE_SOURCE_DIR}/${PALABOS_DIR}/externalLibraries/tinyxml/tinyxml.cpp)
-LIST(APPEND SRC_FILES ${CMAKE_SOURCE_DIR}/${PALABOS_DIR}/externalLibraries/tinyxml/tinyxmlerror.cpp)
-LIST(APPEND SRC_FILES ${CMAKE_SOURCE_DIR}/${PALABOS_DIR}/externalLibraries/tinyxml/tinyxmlparser.cpp)
 
 ADD_library("${PROJECT_NAME}_pre_all_deps" STATIC ${SRC_FILES})
 
@@ -160,5 +151,5 @@ ENDIF(ENABLE_PARMETIS)
 #finalize library
 add_custom_target(hemocell_finalize ALL 
   COMMAND ${CMAKE_AR} crsT lib${PROJECT_NAME}.a
-  $<TARGET_FILE:${PROJECT_NAME}_pre_all_deps>
+  $<TARGET_FILE:${PROJECT_NAME}_pre_all_deps> libhemocell_local.a
   DEPENDS hemocell_pre_finalize hemocell_with_parmetis)
