diff --git a/build/hemocell/CMakeLists.txt b/build/hemocell/CMakeLists.txt
index 915ea22..a6e15c9 100644
--- a/build/hemocell/CMakeLists.txt
+++ b/build/hemocell/CMakeLists.txt
@@ -36,7 +36,7 @@ INCLUDE_DIRECTORIES(${MPI_INCLUDE_PATH})
 #IF (NOT CMAKE_CXX_FLAGS)
   #If flags are not set, add default flags
 #  SET(CMAKE_BUILD_TYPE ${BUILD_TYPE})
-  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3 -march=native -std=c++11")
+  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3 -mtune=generic -std=c++11")
   SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-declarations")
   SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-empty-body")
   SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unknown-pragmas")
